api/include/opentelemetry/nostd/list.h

```diff
@@ -0,0 +1,254 @@
+
+  // Custom iterator class
+  class iterator
+  {
+  public:
+    iterator operator++()
+    {
+      ptr_++;
+      return ptr_;
+    }
+
+    T &operator*() { return *ptr_; }
+
+    T *operator->() { return ptr_; }
+
+    bool operator==(const iterator &other) { return ptr_ == other.ptr_; }
+
+    bool operator!=(const iterator &other) { return ptr_ != other.ptr_; }
+
+  private:
+    friend class list;
+
+    iterator(T *ptr) : ptr_(ptr) {}
+
+    T *ptr_;
+  };
+
+  list()
+  {
+    size_ = 0;
+    data_ = nullptr;
+  };
+
+  // Constructs class with passed in element.
+  list(T element)
+  {
+    size_ = 1;
+
+    data_ = new T[size_];
+
+    std::memcpy(data_, &element, sizeof(T));
+  };
+
+  // Constructs class with the passed in initializer_list.
+  list(std::initializer_list<T> init_list)
+  {
+    size_ = init_list.size();
+
+    data_ = new T[size_];
+
+    std::memcpy(data_, std::begin(init_list), init_list.size() * sizeof(T));
+  }
+
+  // Returns the element at the front of the list.
+  T &front() { return data_[0]; }
+
+  // Returns the element at the back of the list.
+  T &back()
+  {
+    if (size_ == 0)
+    {
+      return data_[0];
+    }
+
+    return data_[size_ - 1];
+  }
+
+  // Returns an iterator pointing at the front of the list.
+  iterator begin() { return iterator(data_); }
+
+  // Returns an iterator pointing at the back of the list.
+  iterator end() { return iterator(data_ + size_); }
+
+  // Inserts the passed in element at the passed in
+  // iterator position.
+  iterator insert(iterator position, const T &element)
+  {
+    int previous_size = size_;
+    size_++;
+
+    T *temp = new T[size_];
+
+    // Counter tracks the index of the passed in iterator position.
+    int counter = position.ptr_ - data_;
+
+    int insert_index       = counter;
+    int next_index         = counter + 1;
+    int remaining_elements = previous_size - counter;
+
+    // Copy the original data that comes before the insertion.
+    std::memcpy(temp, data_, insert_index * sizeof(T));
+    // Copy the inserted data.
+    std::memcpy(temp + insert_index, &element, sizeof(T));
+    // Copy the original data that comes after the insertion.
+    std::memcpy(temp + next_index, data_ + counter, remaining_elements * sizeof(T));
+
+    delete[] data_;
+
+    data_ = temp;
+
+    // Return an iterator at the inserted element.
+    return iterator(data_ + counter);
+  }
+
+  // Insert an element at the end of the list.
+  void push_back(T element)
+  {
+
+    int previous_size = size_;
+    size_++;
+
+    T *temp = new T[size_];
+
+    // Copy the original data into the new array.
+    for (int i = 0; i < previous_size; i++)
+    {
+      temp[i] = data_[i];
+    }
+
+    // Deallocate memory used by previous array.
+    delete[] data_;
+
+    data_ = temp;
+
+    data_[previous_size] = element;
+  };
+
+  // Insert an element at the front of the list.
+  void push_front(T element)
+  {
+    size_++;
+
+    T *temp = new T[size_];
```

Comment:
We might want to store both a `size` and a `capacity`. In this way we could avoid the necessity of `new` and `delete` calls for each list manipulation (which is very costly).

Reply:
Added a capacity member and resize function.